links/endpoint_veth.go

```diff
@@ -14,3 +16,11 @@ func NewEndpointVeth(eg *EndpointGeneric) *EndpointVeth {
 func (e *EndpointVeth) Verify(_ *VerifyLinkParams) error {
 	return CheckEndpointUniqueness(e)
 }
+
+func (e *EndpointVeth) Deploy(ctx context.Context) error {
+	return e.GetLink().Deploy(ctx, e)
+}
+
```

Comment:
this seems weird... We have (endpoint).Deploy(), which in its turn calls (link).Deploy.

Why don't we remove all (endpoint).Deploy and instead call (link).Deploy directly?
The link interface has `GetEndpoints` to access associated endpoints

This is the main thing that makes me suspicious here

Reply:
to expand on this.
Logically we have the following composition of objects:

* Link
  * EndpointA
  * EndpointA

Thus it makes sense that the calling node has `(node).DeployLink` (this func still exists) - https://github.com/srl-labs/containerlab/pull/1914/files#diff-d25cfdb9022b8a280a3cf87aa5a83926d393e4d5d28c550dd57cca87c9130359R519

What I propose to reconsider is what happens next. I think the default node should be 

```
for _, l := range d.Links {
		err := l.Deploy(ctx)
```

since we deploy a link object. Now how and what is needed to deploy this link is with the link implementation. If a link needs to deploy endpoints separately -- it should happen in the `(link).Deploy()`.

That way we keep the logical wrapping intact and

Reply:
> @steiler this seems weird... We have (endpoint).Deploy(), which in its turn calls (link).Deploy.
> 
> Why don't we remove all (endpoint).Deploy and instead call (link).Deploy directly? The link interface has `GetEndpoints` to access associated endpoints
> 
> This is the main thing that makes me suspicious here

The abstraction of having a link attached to a node simply is wrong. The Endpoint belongs to a Node. Regarding the call to Link.Deploy() you see we provide the endpoint as parameter. This is, because it matters which side (Endpoint) of the Link calls the creation. Sure we could search through the list via provided node names bla bla bla ... still it would be the wrong abstraction of the whole thing.

Reply:
> to expand on this. Logically we have the following composition of objects:
> 
> * Link
>   
>   * EndpointA
>   * EndpointA
> 
> Thus it makes sense that the calling node has `(node).DeployLink` (this func still exists) - https://github.com/srl-labs/containerlab/pull/1914/files#diff-d25cfdb9022b8a280a3cf87aa5a83926d393e4d5d28c550dd57cca87c9130359R519
> 
> What I propose to reconsider is what happens next. I think the default node should be
> 
> ```
> for _, l := range d.Links {
> 		err := l.Deploy(ctx)
> ```
> 
> since we deploy a link object. Now how and what is needed to deploy this link is with the link implementation. If a link needs to deploy endpoints separately -- it should happen in the `(link).Deploy()`.
> 
> That way we keep the logical wrapping intact and

Again, the node itself needs to specifically know about its side of the link, namely the Endpoint. It contains the name that a link is referenced with etc. So the endpoint belongs to the Node and the Endpoint is part of a Link. Just like that.
We've had these discussions last time and I surrendered, but figured again this time that it was bullshit.
What we did so far was to keep a list of links and endpoints. Which broke my neck because I was not keeping these lists in sync... avoidable by adding just endpoints.
If you where just to store links, whenever you would want to access the Endpoint via the link, you would need to resolve it via a loop, looking up the node name, which is stupid.
So I remain with my opinion that this is the right abstraction for the whole matter.

Reply:
> Thus it makes sense that the calling node has `(node).DeployLink` (this func still exists) - https://github.com/srl-labs/containerlab/pull/1914/files#diff-d25cfdb9022b8a280a3cf87aa5a83926d393e4d5d28c550dd57cca87c9130359R519

Thanks for spotting this, I'll rename it to (node).EndpointsDeploy()